rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileMetadata.java

```diff
@@ -55,15 +58,29 @@ public FileMetadata()
   {
   }
 
-  public FileMetadata(String remotePath, long fileLength, long lastModified, long currentFileSize, Configuration conf)
+  public FileMetadata(String remotePath, long fileLength, long lastModified, long currentFileSize, Configuration conf, boolean updateMetadata)
       throws IOException
   {
     this.remotePath = remotePath;
     this.size = fileLength;
     this.lastModified = lastModified;
     this.currentFileSize = currentFileSize;
-    localPath = CacheUtil.getLocalPath(remotePath, conf);
-    mdFilePath = CacheUtil.getMetadataFilePath(remotePath, conf);
+    if (!updateMetadata) {
+      if (generationNumber.containsKey(remotePath)) {
+        genNumber = generationNumber.get(remotePath);
+        genNumber++;
+        while (Files.exists(Paths.get(CacheUtil.getMetadataFilePath(remotePath, conf, genNumber)))) {
+          genNumber++;
```

Comment:
You need to also consider the mode where it is expected to resume from the state on disk if rubix servers had crashed. This is tricky because we want all three things:
1. No memory leaks on generationNumber map
2. Support resuming from on-disk state
3. Also skip generationNumbers for which files already exist when the case is not of resuming from on-disk state